### DISCUSSION TITLE
`QF_FF` and binary decompositions?

### DISCUSSION THREAD

**Discussion by gbotrel** [cvc5/cvc5#9532] [Q&A] [answered]
Hi! Doing some experiments to use `QF_FF` with [gnark](https://github.com/consensys/gnark) library. 
I'm hitting a perf bottleneck and looking for suggestions on how to better use `cvc` / encode the problem; say we want to decompose a finite field element `e` in `n` bits; 
1. we introduce `n` variables that we constrain to be `0` or `1`
2. we constrain the sum of the bits with the coefficients (powers of 2) to be equal to `e` : `Σ (2**i * b[i]) == e` 

With `n` small (< 8) `cvc` solver finds sat / unsat in a reasonable time. When `n` grows (to 128 or 256 bits for example), it does not. 

Question 1: is there a way to give a "hint" to the `cvc` solver on how to assign values to these `n`-bits variables rather than letting it explore all the paths? 
Question 2: would it perform better to mix `QF_FF` with another theory?

Here is a snippet I'm running, with `n=4`;

```cpp

#include <cvc5/cvc5.h>
#include "error.H"
#include <cassert>
#include <iostream>

using namespace std;
using namespace cvc5;

const string fModulus =
    "21888242871839275222246405745257275088548364400416034343698204186575808495617";

#define variable(v) solver.mkConst(field, #v)
#define constant(v) solver.mkFiniteFieldElem(#v, field)
#define add(...) solver.mkTerm(FINITE_FIELD_ADD, {__VA_ARGS__})
#define mul(...) solver.mkTerm(FINITE_FIELD_MULT, {__VA_ARGS__})
#define neg(...) solver.mkTerm(FINITE_FIELD_NEG, {__VA_ARGS__})
#define eq(...) solver.mkTerm(EQUAL, {__VA_ARGS__})
// #define print(v) solver.getValue(v).getFiniteFieldValue()
#define print(v)                                                                                   \
    cout << #v << " = "                                                                            \
         << solver.getValue(v).toString().substr(2, solver.getValue(v).toString().length() -       \
                                                        fModulus.length() - 3)                     \
         << endl

// build command: g++ -std=c++20 main.cpp -lcvc5 -o main
int main() {
    Solver solver;
    solver.setOption("produce-models", "true");
    solver.setLogic("QF_FF");

    Sort field = solver.mkFiniteFieldSort(fModulus);
    // public and secret variables
    auto X = variable(X);
    auto one = constant(1);

    // internal variables
    auto v0 = variable(v0);
    auto v1 = variable(v1);
    auto v2 = variable(v2);
    auto v3 = variable(v3);
    auto v4 = variable(v4);

    // coefficients
    auto c0 = constant(0);
    auto c1 = constant(1);
    auto c2 = constant(2);
    auto c3 = constant(4);
    auto c4 = constant(8);

    // constraint 0: v0 ⋅ 1 + -1⋅v0 == 0
    solver.assertFormula(eq(v0, c0).orTerm(eq(v0, c1)));

    // constraint 1: v1 ⋅ 1 + -1⋅v1 == 0
    solver.assertFormula(eq(v1, c0).orTerm(eq(v1, c1)));

    // constraint 2: v2 ⋅ 1 + -1⋅v2 == 0
    solver.assertFormula(eq(v2, c0).orTerm(eq(v2, c1)));

    // constraint 3: v3 ⋅ 1 + -1⋅v3 == 0
    solver.assertFormula(eq(v3, c0).orTerm(eq(v3, c1)));

    // constraint 4: 1 ⋅ v0 + 2⋅v1 + 4⋅v2 + 8⋅v3 == X
    solver.assertFormula(eq(add(v0, mul(c2, v1), mul(c3, v2), mul(c4, v3)), X));

    // omitting constraints that should be unsat [...]
     
    // check sat

    try {
        Result r = solver.checkSat();
        if (r.isNull() || r.isUnknown()) {
            cout << "unknown result" << endl;
        } else {
            if (r.isSat()) {
                cout << "❌ satisfied! ❌" << endl;
                print(X);

                return -1;
            } else {
                cout << "unsat ✅" << endl;
                return 0;
            }
        }
    } catch (CoCoA::ErrorInfo e) {
        cout << e << endl;
        return -1;
    }
}
```

**Comment by gbotrel** [comment]
cc @alex-ozdemir :)

**Comment by alex-ozdemir** [comment, marked as answer]
Hi Gautam,

First, thank you for your work on gnark. I know some folks who use (and love!) it.

> Question 1: is there a way to give a "hint" to the cvc solver on how to assign values to these n-bits variables rather than letting it explore all the paths?

Right now, the field solver can't use information like this. But, it should be possible to add this kind of thing, given the architecture of our solver. I've been somewhat interested in this for a while, but haven't implemented it yet.

>  Question 2: would it perform better to mix QF_FF with another theory?

I don't think so.

Out of curiosity, what are you using cvc5's FF solver for?

**Reply by gbotrel** [reply]
heh, mostly exploring (solo) at the moment; the three more or less independent ideas I have are;
1. adding some sort of minimalist "pre/post conditions" annotations to `gnark` API. as a first step I did some basic validations of the core APIs using cvc to prove some scenarios to be unsat, as expected. 
2. since we support multiple frontend, with possibly multiple type of constraint systems; use `cvc` to detect non-equivalence of constraint generated by/for different frontend/constraint system types. Being able to say "using `gnark` frontend with these options to generate degree 5 constrains, generates `n` constraints which are equivalent to these `m` constraints of degree 2 would be a useful tool (internally). 
3. make a tool to help identify bugs / generate test vectors. it's not uncommon to have someone ping us with "hey I wrote this circuit but I have this error"

But bit decompositions are really ubiquitous; so glad to know there is a way to go around this blocker :) . 

Can you give me some pointers onto where cvc solver makes the decision to assign a value to a var and / or some rough idea on what you had in mind? Might be able to do a "quick and dirty" version that would unblock this 👍

**Reply by alex-ozdemir** [reply]
>     1. adding some sort of minimalist "pre/post conditions" annotations to `gnark` API. as a first step I did some basic validations of the core APIs using cvc to prove some scenarios to be unsat, as expected.
> 
>     2. since we support multiple frontend, with possibly multiple type of constraint systems; use `cvc` to detect non-equivalence of constraint generated by/for different frontend/constraint system types. Being able to say "using `gnark` frontend with these options to generate degree 5 constrains, generates `n` constraints which are equivalent to these `m` constraints of degree 2 would be a useful tool (internally).
> 
>     3. make a tool to help identify bugs / generate test vectors. it's not uncommon to have someone ping us with "hey I wrote this circuit but I have this error"

Great! All of these ideas make sense!

> But bit decompositions are really ubiquitous; so glad to know there is a way to go around this blocker :) .
> 
> Can you give me some pointers onto where cvc solver makes the decision to assign a value to a var and / or some rough idea on what you had in mind? Might be able to do a "quick and dirty" version that would unblock this +1

I'll give you some pointers. But, no promises about there being a quick
hack---you may still be blocked.

The FF-solver's internal search is in this file[1]. It decides how to branch
in this function[2]. You might be able to express your type of reasoning as a
different way of branching (or, perhaps NOT branching).

[1]: https://github.com/cvc5/cvc5/blob/main/src/theory/ff/multi_roots.cpp
[2]: https://github.com/cvc5/cvc5/blob/main/src/theory/ff/multi_roots.cpp#L156

I'm not very sure that this kind of hack would work that well though.
Something higher-level, with propagations, might be better. Or perhaps some
conversion hack b/w the BV theory and the FF theory.

If I didn't make this clear before: being efficient for this kind of reasoning
is possible, but not easy. It may be a research problem.
